core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -329,7 +330,14 @@ protected void validateJoinFilters(List<RexNode> aboveFilters,
     final Iterator<RexNode> filterIter = joinFilters.iterator();
     while (filterIter.hasNext()) {
       RexNode exp = filterIter.next();
-      if (!predicate.apply(join, joinType, exp)) {
+      int maxIndex = -1;
+      ImmutableList<RexInputRef> inputRefs = RexUtil.getAllInputRef(exp);
+      for (RexInputRef rexInputRef: inputRefs) {
+        if (rexInputRef.getIndex() > maxIndex) {
+          maxIndex = rexInputRef.getIndex();
+        }
+      }
+      if (!predicate.apply(join, joinType, exp) && maxIndex < join.getRowType().getFieldCount()) {
```

Comment:
Well, I'm surprised to see this -- Given `joinFilters` is from join condition, how can it contains a `RexInputRef` which has index bigger than `join.getRowType().getFieldCount()` ?

Reply:
It's the filter in join condition, from
`final List<RexNode> joinFilters =
        RelOptUtil.conjunctions(join.getCondition());`
For the case in jira
```
EnumerableFilter.ENUMERABLE.[](input=RelSubset#38,condition==($1, 0))
  EnumerableHashJoin.ENUMERABLE.[](left=RelSubset#17,right=RelSubset#73,condition==($0, $3),joinType=semi)
    EnumerableTableScan(subset=[rel#17:Subset#0.ENUMERABLE.[]], table=[[tblspace1, tsql]])
    LogicalAggregate(subset=[rel#23:Subset#3.NONE.[]], group=[{0}])
      LogicalProject(subset=[rel#21:Subset#2.NONE.[]], K1=[$0])
        LogicalFilter(subset=[rel#19:Subset#1.NONE.[]], condition=[=($1, 1)])
          EnumerableTableScan(subset=[rel#17:Subset#0.ENUMERABLE.[]], table=[[tblspace1, tsql]])
```
it is `==($0, $3)`. For semi join, it might bigger than field count of join.

Reply:
Ok, I got it ~